versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,376 @@
+    this.config = config;
+    this.collections = new HashMap<>();
+    final CodecRegistry codecRegistry = CodecRegistries.fromProviders(
+        new CodecProvider(),
+        PojoCodecProvider.builder().automatic(true).build(),
+        MongoClientSettings.getDefaultCodecRegistry());
+    this.mongoClientSettings = MongoClientSettings.builder()
+      .applyConnectionString(new ConnectionString(config.getConnectionString()))
+      .codecRegistry(codecRegistry)
+      .writeConcern(WriteConcern.MAJORITY)
+      .build();
+  }
+
+  /**
+   * Gets a handle to an existing database or get a handle to a MongoDatabase instance if it does not exist. The new
+   * database will be lazily created.
+   * Since MongoDB creates databases and collections if they do not exist, there is no need to validate the presence of
+   * either before they are used. This creates or retrieves collections that map 1:1 to the enumerates in
+   * {@link com.dremio.nessie.versioned.store.ValueType}
+   */
+  @Override
+  public void start() {
+    mongoClient = MongoClients.create(mongoClientSettings);
+    mongoDatabase = mongoClient.getDatabase(config.getDatabaseName());
+
+    // Initialise collections for each ValueType.
+    collections.put(ValueType.L1, mongoDatabase.getCollection(config.getL1TableName(), L1.class));
+    collections.put(ValueType.L2, mongoDatabase.getCollection(config.getL2TableName(), L2.class));
+    collections.put(ValueType.L3, mongoDatabase.getCollection(config.getL3TableName(), L3.class));
+    collections.put(ValueType.COMMIT_METADATA,
+        mongoDatabase.getCollection(config.getMetadataTableName(), InternalCommitMetadata.class));
+    collections.put(ValueType.KEY_FRAGMENT,
+        mongoDatabase.getCollection(config.getKeyListTableName(), Fragment.class));
+    collections.put(ValueType.REF, mongoDatabase.getCollection(config.getRefTableName(), InternalRef.class));
+    collections.put(ValueType.VALUE,
+        mongoDatabase.getCollection(config.getValueTableName(), InternalValue.class));
+  }
+
+  /**
+   * Closes the connection this manager creates to a database. If the connection is already closed this method has
+   * no effect.
+   */
+  @Override
+  public void close() {
+    if (null != mongoClient) {
+      mongoClient.close();
+    }
+  }
+
+  @Override
+  public void load(LoadStep loadstep) throws ReferenceNotFoundException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public <V> boolean putIfAbsent(ValueType type, V value) {
+    final MongoCollection<V> collection = (MongoCollection<V>)collections.get(type);
+    if (null == collection) {
+      throw new UnsupportedOperationException(String.format("Unsupported Entity type: %s", type.name()));
+    }
+
+    final UpdateResult result = await(() -> collection.updateOne(
+        Filters.eq(Store.KEY_NAME, ((HasId)value).getId()),
+        new UpdateEntityBson<>(value),
+        UPDATE_UPSERT)).first();
```

Comment:
This doesn't look right. Doesn't upsert patch the object if it is already there? If so, that is not what we want.